lib/experimental/interactivity-api/class-wp-interactivity-store.php

```diff
@@ -67,7 +57,9 @@ static function render() {
 		if ( empty( self::$store ) ) {
 			return;
 		}
-		$store = self::serialize();
-		echo "<script id=\"wp-interactivity-store-data\" type=\"application/json\">$store</script>";
+		echo sprintf(
+			'<script id="wp-interactivity-store-data" type="application/json">%s</script>',
+			wp_json_encode( self::$store )
```

Comment:
here we ought to also escape potentially ambiguous characters

```php
wp_json_encode( self::$store, JSON_HEX_TAG | JSON_HEX_AMP )
```

Reply:
For context (Dennis explained this to me privately), a user could inject `{"content": "this is a </script>"}` and it will close the script element unexpectedly. Or a malicious user could inject `{"content": "this is a </script><script>malicious();</script>"}` and execute malicious code.

Using `JSON_HEX_TAG` turns that into `{"content": "this is a \u003C/script\u003E"}`.

Reply:
I wanted to add a small PHPUnit test for this, and it turned out a malicious script can't be injected that way because **trailing slashes are escaped by default** (there is a flag to deactivate this behavior, see [`JSON_UNESCAPED_SLASHES`](https://www.php.net/manual/en/json.constants.php#:~:text=to%20format%20it.-,JSON_UNESCAPED_SLASHES,-(int))). 🤷

The example above actually becomes as follows:
```
{"content":"this is a <\/script><script>malicious();<\/script>"}
```

As `<\/script>` is not a valid closing tag, the `<script>` remains open, and the malicious code cannot be interpreted and executed.

However, this already happens with trailing slashes anywhere! E.g.,
```json
{ "url": "http://mydomain.com/page/123" }
```

turns into
```
{"url":"http:\/\/mydomain.com\/page\/123"}
```

So `json_encode` seems to be _secure_ by default (in that regard), although it is modifying strings in a way that we are not handling correctly in the Interactivity API runtime.

Now, my questions:
- Should we stay with the json_encode's default behavior (secure) and, in the client, process the `<script>` content to convert all `\/` to `/` and then parse the result?
- Or should we use the above flags, along with `JSON_UNESCAPED_SLASHES`, to avoid processing the `<script>` content in the client _anyway_?

Reply:
Interesting.

> process the <script> content to convert all \/ to / and then parse the result?

`JSON.parse` seems to be handling those slashes well, including the dangerous `script` tag:

```js
JSON.parse('{"url":"http:\/\/mydomain.com\/page\/123"}').url
// outputs 'http://mydomain.com/page/123'

JSON.parse('{"content":"this is not a <\/span><span>malicious<\/span> content"}').content
// outputs 'this is not </span><span>malicious</span> content'

JSON.parse('{"content":"this is a <\/script><script>malicious();<\/script> script"}').content
// outputs 'this is a \x3C/script>\x3Cscript>malicious();\x3C/script> script'
```

So I don't think we need to do anything in terms of parsing.

Reply:
Thanks, David. Let's wait until we hear Dennis' opinion about this 🙂

Reply:
this seems mostly fine; thanks for testing everything. I've tried myself and can't figure out how to break it the way I thought I could. I've asked for external review.

Reply:
Thanks, Dennis! 

> I've asked for external review.

Should we wait?

Reply:
I think it's wise to be cautious :) There's no penalty to using `JSON_HEX_TAG` other than making the raw HTML slightly harder for a human to read.

I know of one way the current output (without the `JSON_HEX_TAG` flag) can be broken: an opening comment tag.

What content the HTML spec allows inside `<script>` tags (even those that have a non-executable `type` attribute value) is [really odd](https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements).

Try saving the following as `wat.php` and loading it in a browser:
```
<!DOCTYPE html>
<html>
JavaScript is weird
<!-- First Script -->
<script type="application/json"><?php echo json_encode( ["one" => "two<!-- <script three"] ); ?></script>
<!-- Second Script -->
<script>document.write('but fun')</script>
right?
</html>
```

At first glance, it looks like it should generate a DOM that, to a human, would read "JavaScript is weird but fun right?". Instead, you'll see that it reads "JavaScript is weird right?". The problem is the `<!-- <script` in the first script. When a browser sees a byte sequence like that inside a `<script>` tag, it will keep that script tag open until it finds a closing comment tag (`-->`) followed by a closing script tag (`</script>`) (with anything between the two).

In the example above, that rule means that the browser only "sees" one script tag, and it has the following contents:
```
<?php echo json_encode( ["one" => "two<!-- <script three"] ); ?></script>
<!-- Second Script -->
<script>document.write('but fun')
```

(Viewing the page source of wat.php in your browser or inspecting the DOM will make this clear.)

I don't know if it's possible for a byte sequence like `<!-- <script` to appear in the data being JSON encoded here, but let's do
```
wp_json_encode( self::$store, JSON_HEX_TAG | JSON_HEX_AMP )
```
as suggested anyway. `JSON_HEX_TAG` will prevent any of the above weirdness. `JSON_HEX_AMP` isn't at all related, but it may prevent some further weirdness if anyone is trying to serve these pages as XHTML (a very very odd thing to do these days :)). There may be a couple other odd edge cases where `JSON_HEX_AMP` is handy too - I don't remember (though they probably only come up in old browsers).

Reply:
I did confirm that using `echo json_encode( [ 'one' => '<!-- <script' ] )` results in the unwanted output to the browser. thanks for remembering that case, @mdawaffe

Reply:
Wow, thank you very much for the explanation, @mdawaffe! 🙏  And also you, @dmsnell, for suggesting that change. 😄